botocore/utils.py

```diff
@@ -919,6 +956,8 @@ def is_valid_endpoint_url(endpoint_url):
         re.IGNORECASE)
     return allowed.match(hostname)
 
+def is_valid_ipv4_or_ipv6_endpoint(endpoint_url):
+    return is_valid_endpoint_url(endpoint_url) or is_valid_IPv6_endpoint_url(endpoint_url)
```

Comment:
I think we have an edge case here where unbracketed shorthand ipv6 addresses will pass the ipv4 check. Those aren't valid URIs, so we'll need to handle them appropriately. We either need to add additional constraints here to use the ipv6 check on an address with more than one `:` (we may need to evaluate without user information e.g. `url.netloc.split('@')[-1]`).

Alternatively, we could use urllib3's validation but need to scope it more appropriately as discussed above.